### PR TITLE
Implementation of pseudocost observations.

### PR DESCRIPTION
**Pull request by dchetelat**
## Pull request checklist

- [x] I have open an issue to discuss the proposed changes. Fix #
- [x] I have modified/added tests to cover the new changes/features.
- [ ] I have modified/added the documenation to cover the new changes/features.
- [x] I have ran the tests, checks, and code formatters.

## Proposed implementation
A new `ObservationFunction` that returns pseudocosts. Analoguous to the current `StrongBranchingScores`

### CODE REVIEW

**Comment by dchetelat** [comment]
@AntoinePrv What do I need to do to pass the new "documentation" check?

File: libecole/include/ecole/observation/pseudocosts.hpp [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+#include <xtensor/xview.hpp>
+
+#include "ecole/observation/abstract.hpp"
+
+namespace ecole {
+namespace observation {
+
+class Pseudocosts : public ObservationFunction<nonstd::optional<xt::xtensor<double, 1>>> {
+public:
+	nonstd::optional<xt::xtensor<double, 1>> obtain_observation(scip::Model& state) override;
```
Comment (AntoinePrv) [inline libecole/include/ecole/observation/pseudocosts.hpp:16]: Can we call that variable `model`

File: libecole/include/ecole/observation/pseudocosts.hpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include <memory>
+
+#include <nonstd/optional.hpp>
+#include <xtensor/xtensor.hpp>
+#include <xtensor/xview.hpp>
```
Comment (AntoinePrv) [inline libecole/include/ecole/observation/pseudocosts.hpp:7]: I don't believe we need that include here. If it is needed in the cpp file, please move it there.

File: libecole/src/observation/pseudocosts.cpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		int nb_cands;
+
+		/* get branching candidates */
+		SCIPgetLPBranchCands(scip, &cands, &cands_lp_values, NULL, NULL, &nb_cands, NULL);
+		assert(nb_cands > 0);
+
+		/* Store pseudocosts in tensor */
+		auto const nb_lp_columns = static_cast<std::size_t>(SCIPgetNLPCols(scip));
+		auto pseudocosts = xt::xarray<double>::from_shape({nb_lp_columns});
+		pseudocosts.fill(std::nan(""));
```
Comment (AntoinePrv) [inline libecole/src/observation/pseudocosts.cpp:32]: - I believe it should be a `xt::xtensor`
- I believe you can use in one shot initialization and filling with
   ```cpp
   xt::xtensor<double, 1> pseudocosts({nb_lp_columns}, std::nan(""));
   ```

File: libecole/src/observation/pseudocosts.cpp [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+		SCIP_COL* col;
+		SCIP_Real score;
+		int lp_index;
```
Comment (AntoinePrv) [inline libecole/src/observation/pseudocosts.cpp:36]: Variable in the loop don't need to be declared before the loop...

File: libecole/src/observation/pseudocosts.cpp [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+		for (int i = 0; i < nb_cands; i++) {
+			col = SCIPvarGetCol(cands[i]);
+			lp_index = SCIPcolGetLPPos(col);
+			score = SCIPgetVarPseudocostScore(scip, cands[i], cands_lp_values[i]);
```
Comment (AntoinePrv) [inline libecole/src/observation/pseudocosts.cpp:40]: ... simply add `auto` before these variables.

File: libecole/tests/src/observation/test-pseudocosts.cpp [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#include "ecole/environment/branching.hpp"
+#include "ecole/environment/exception.hpp"
+#include "ecole/observation/pseudocosts.hpp"
+#include "ecole/reward/isdone.hpp"
+
+#include "conftest.hpp"
+
+using namespace ecole;
+
+TEST_CASE("Pseudocosts") {
```
Comment (AntoinePrv) [inline libecole/tests/src/observation/test-pseudocosts.cpp:17]: So unittests for observations are a bit of a mess right now, and I'll work on making them better.
Those tests here show that the interface works correctly with Ecole.

I'd like to see a functional test, something that says, "_Hey look, my function is not returning a vector of random number, this is a correct output_".
I understand it is hard to predict what the pseudo cost will be, so maybe tests that

- The vector has as many entries as there are variables;
- The vector is not filled only with `nan`
- Pseudo cost are positive (for minimization problems but not sure of SCIP conventions here)
- If a variable is fractional, it has a pseudo cost
- ...

File: python/src/ecole/core/observation.cpp [hunk 7]
```diff
@@ -146,6 +147,24 @@ void bind_submodule(py::module m) {
...
+
+	auto pseudocosts = py::class_<Pseudocosts>(m, "Pseudocosts", R"(
+		Pseudocosts observation function on branch-and bound node.
+
+		This observation obtains pseudocosts for all LP fractional candidate variables at a
+		branch-and-bound node.  The pseudocost is a cheap approximation to the strong branching
+        score and measures the quality of branching for each variable.  This observation can be used
+        as a practical branching strategy by always branching on the variable with the highest
+        pseudocost, although in practice is it not as efficient as SCIP's default strategy, reliability
+        pseudocost branching (also known as hybrid branching).
```
Comment (AntoinePrv) [inline python/src/ecole/core/observation.cpp:159]: Weird indentation

File: python/src/ecole/core/observation.cpp [hunk 8]
```diff
@@ -146,6 +147,24 @@ void bind_submodule(py::module m) {
...
+
+		This observation function extracts an array containing the pseudocost for
+		each variable in the problem which can be indexed by the action set.  Variables for which
+		a pseudocost is not applicable are filled with NaN.
+	)");
+	pseudocosts.def(py::init<>());
+	def_reset(pseudocosts, "Cache some feature not expected to change during an episode.");
```
Comment (AntoinePrv) [inline python/src/ecole/core/observation.cpp:166]: Not is does not. Just put "Do nothing"

**Comment by AntoinePrv** [comment]
> @AntoinePrv What do I need to do to pass the new "documentation" check?

You can add the class under `/docs/reference/observation.rst`, section `Listing`.

**Comment by dchetelat** [comment]
> > @AntoinePrv What do I need to do to pass the new "documentation" check?
> 
> You can add the class under `/docs/reference/observation.rst`, section `Listing`.

I tried changing this file for
```
Observations
============

Interface
---------


Listing
-------
Node Bipartite
^^^^^^^^^^^^^^
.. autoclass:: ecole.observation.NodeBipartite
   :members:
.. autoclass:: ecole.observation.NodeBipartiteObs
   :members:

Pseudocosts
^^^^^^^^^^^
.. autoclass:: ecole.observation.Pseudocosts
   :members:

Utilities
---------
Nothing
^^^^^^^
.. autoclass:: ecole.observation.Nothing
   :members:

Tuple of Observations
^^^^^^^^^^^^^^^^^^^^^
.. autoclass:: ecole.observation.TupleFunction
   :members:

Dictionnary of Observations
^^^^^^^^^^^^^^^^^^^^^^^^^^^
.. autoclass:: ecole.observation.DictFunction
   :members:
~                     
```

but I get the error
```
(ecole) chetdidi@nero.gerad.lan:/local1/chetdidi/ecole>cmake --build build/ --target ecole-sphinx
[  3%] Parsing C++ documentation with Doxygen
[  3%] Built target ecole-doxygen
[ 65%] Built target libecole
[ 88%] Built target ecole-python
[ 92%] Installing ecole in virtual environment
[ 96%] Built target ecole-venv
[100%] Generating documentation with Sphinx

Warning, treated as error:
docstring of ecole.observation.Pseudocosts:5:Block quote ends without a blank line; unexpected unindent.
gmake[3]: *** [docs/CMakeFiles/ecole-sphinx] Erreur 2
gmake[2]: *** [docs/CMakeFiles/ecole-sphinx.dir/all] Erreur 2
gmake[1]: *** [docs/CMakeFiles/ecole-sphinx.dir/rule] Erreur 2
gmake: *** [ecole-sphinx] Erreur 2
```

Moreover you haven't told me where I am actually supposed to add the actual documentation, and what format it should look like.

File: libecole/include/ecole/observation/pseudocosts.hpp (hunk 1 above)
Comment (dchetelat) [reply libecole/include/ecole/observation/pseudocosts.hpp:16]: Yes that's a typo, good catch

File: libecole/include/ecole/observation/pseudocosts.hpp (hunk 2 above)
Comment (dchetelat) [reply libecole/include/ecole/observation/pseudocosts.hpp:7]: Done

File: libecole/src/observation/pseudocosts.cpp (hunk 3 above)
Comment (dchetelat) [reply libecole/src/observation/pseudocosts.cpp:32]: This is from
https://github.com/ds4dm/ecole/blob/727af527279a23e42c8f061fccdc0465a390523e/libecole/src/observation/strongbranchingscores.cpp#L68
which code you previously accepted and merged into master.

File: libecole/src/observation/pseudocosts.cpp (hunk 4 above)
Comment (dchetelat) [reply libecole/src/observation/pseudocosts.cpp:36]: I don't understand what you mean here.

File: libecole/src/observation/pseudocosts.cpp (hunk 5 above)
Comment (dchetelat) [reply libecole/src/observation/pseudocosts.cpp:40]: What does `auto` do? Why do I need it?

File: python/src/ecole/core/observation.cpp (hunk 7 above)
Comment (dchetelat) [reply python/src/ecole/core/observation.cpp:159]: Fixed, I had spaces rather than tabs.

File: python/src/ecole/core/observation.cpp (hunk 8 above)
Comment (dchetelat) [reply python/src/ecole/core/observation.cpp:166]: I got this from
https://github.com/ds4dm/ecole/blob/71f3c6d544512642b1c032956befe3a10df09e08/python/src/ecole/core/observation.cpp#L146-L147
and I don't think the reset there does anything either?

File: libecole/include/ecole/observation/pseudocosts.hpp (hunk 2 above)
Comment (dchetelat) [reply libecole/include/ecole/observation/pseudocosts.hpp:7]: Hmm looks like `clang-tidy` complains if I don't include them? Yet it compiles fine.
```
clang-tidy...............................................................Failed
- hook id: clang-tidy
- exit code: 1

/local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp:10:48: error: use of undeclared identifier 'nonstd' [clang-diagnostic-error]
class Pseudocosts : public ObservationFunction<nonstd::optional<xt::xtensor<double, 1>>> {
                                               ^
/local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp:10:65: error: use of undeclared identifier 'xt' [clang-diagnostic-error]
class Pseudocosts : public ObservationFunction<nonstd::optional<xt::xtensor<double, 1>>> {
                                                                ^
/local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp:10:88: error: expected a type [clang-diagnostic-error]
class Pseudocosts : public ObservationFunction<nonstd::optional<xt::xtensor<double, 1>>> {
                                                                                       ^
/local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp:12:2: error: use of undeclared identifier 'nonstd' [clang-diagnostic-error]
        nonstd::optional<xt::xtensor<double, 1>> obtain_observation(scip::Model& model) override;
        ^
/local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp:12:19: error: use of undeclared identifier 'xt' [clang-diagnostic-error]
        nonstd::optional<xt::xtensor<double, 1>> obtain_observation(scip::Model& model) override;
                         ^
5 errors generated.
Error while processing /local1/chetdidi/ecole/libecole/include/ecole/observation/pseudocosts.hpp.
```

Comment (AntoinePrv) [reply libecole/include/ecole/observation/pseudocosts.hpp:7]: Well you do need `<nonstd/optional.hpp>` and `<xtensor/xtensor.hpp>`, but I believe not `<xtensor/xview.hpp>`.
Unnecessary headers increase compile time in all files that will include `pseudocost.hpp`.

File: libecole/src/observation/pseudocosts.cpp (hunk 3 above)
Comment (AntoinePrv) [reply libecole/src/observation/pseudocosts.cpp:32]: I understand, but my reviews are not perfect. Every new ones bring new opportunity for improvement.

File: python/src/ecole/core/observation.cpp (hunk 8 above)
Comment (AntoinePrv) [reply python/src/ecole/core/observation.cpp:166]: True, a mistake that should also be corrected.

File: libecole/src/observation/pseudocosts.cpp (hunk 3 above)
Comment (dchetelat) [reply libecole/src/observation/pseudocosts.cpp:32]: Ahah yeah I know :-) I was more saying, we should probably do similar changes in the strong branching code too then. Do you want me to incorporate that (fixes to SB code) in this pull request as well?

File: libecole/include/ecole/observation/pseudocosts.hpp (hunk 2 above)
Comment (dchetelat) [reply libecole/include/ecole/observation/pseudocosts.hpp:7]: Ah ok, good catch! I'll fix that then.

File: libecole/src/observation/pseudocosts.cpp (hunk 3 above)
Comment (AntoinePrv) [reply libecole/src/observation/pseudocosts.cpp:32]: That would be great! Perhaps make it a separate commit though?

File: libecole/src/observation/pseudocosts.cpp (hunk 4 above)
Comment (AntoinePrv) [reply libecole/src/observation/pseudocosts.cpp:35]: Wait, this is not resolved ^^
Let me do a commit for you to see

File: python/src/ecole/core/observation.cpp [hunk 9]
```diff
@@ -142,10 +143,27 @@ void bind_submodule(py::module m) {
...
+		as a practical branching strategy by always branching on the variable with the highest
+		pseudocost, although in practice is it not as efficient as SCIP's default strategy, reliability
+		pseudocost branching (also known as hybrid branching).
+
+		This observation function extracts an array containing the pseudocost for
+		each variable in the problem which can be indexed by the action set.  Variables for which
+		a pseudocost is not applicable are filled with NaN.
+	)");
+	pseudocosts.def(py::init<>());
+	def_reset(pseudocosts, "(Do nothing.)");
```
Comment (AntoinePrv) [inline python/src/ecole/core/observation.cpp:165]: Either use raw string with parenthesis `R"(Do nothing.)"`, or normal string without `"Do nothing."`

**Comment by AntoinePrv** [comment]
I've done the change for the loop variable so that you see what I meant.
I still think tests need improvements,

File: libecole/src/observation/pseudocosts.cpp (hunk 4 above)
Comment (dchetelat) [reply libecole/src/observation/pseudocosts.cpp]: Noooo! don't do that. I have the commit locally, now you just made a mess...

**Comment by dchetelat** [comment]
I strengthened the unit test as you requested.

**Pull request merged**